### PR TITLE
refactor: Document and cleanup unification

### PR DESCRIPTION
**Pull request by Marwes**

### CODE REVIEW

**Comment by Marwes** [comment]
Nothing to exiting here, just some cleanup to make #77 easier and a few comments added.

File: check/src/typecheck.rs [hunk 1]
```diff
@@ -1235,8 +1237,8 @@ impl<'a> Typecheck<'a> {
         debug!("Unify {} <=> {}",
                types::display_type(&self.symbols, expected),
                types::display_type(&self.symbols, &actual));
-        let mut state = ::unify_type::State::new(&self.environment);
-        match unify::unify(&self.subs, &mut state, expected, &actual) {
+        let state = ::unify_type::State::new(&self.environment);
```
Comment (brendanzab) [inline check/src/typecheck.rs]: Maybe import this as `UnifyState` or something?

File: check/src/unify.rs [hunk 2]
```diff
@@ -31,122 +31,133 @@ impl<T, E> fmt::Display for Error<T, E>
...
-        let mut errors = &mut unifier.unifier.errors;
+    fn try_match(unifier: &mut UnifierState<S, Self>, l: &T, r: &T) -> Option<T> {
+        let subs = unifier.unifier.subs;
+        // Retrieve the 'real' types by resolving
         let l = subs.real(l);
         let r = subs.real(r);
-        match (l.get_var(), r.get_var()) {
-            (Some(l), Some(r)) if l.get_id() == r.get_id() => None,
+        // `l` and `r` must have the same type, if one is a variable that variable is
+        // 'unioned' with whatever the other type is
```
Comment (brendanzab) [inline check/src/unify.rs]: unioned -> unified?

File: check/src/typecheck.rs (hunk 1 above)
Comment (Marwes) [inline check/src/typecheck.rs]: There were a bunch of other places with `::unify_type::*` so I added `use unify_type;` at the top instead. `unify_type` is not an ideal name but just using `unify` may confuse it with the `unify` module itself.

File: check/src/unify.rs (hunk 2 above)
Comment (Marwes) [inline check/src/unify.rs]: Fixed

**Pull request merged**